### ISSUE TITLE
Corruption at the end of a WAV stream.

### ISSUE THREAD

**Issue by SiegeLord** [liballeg/allegro5#474]
See https://www.allegro.cc/forums/thread/615461.

**Comment by fatcerberus** [comment]
Any progress on figuring this out?

**Comment by beoran** [comment]
Your help is most welcome :)

**Comment by fatcerberus** [comment]
I actually did look into Allegro's audio code trying to diagnose this a while back, but nothing jumped out at me, everything looked normal.  I guess I'll give it another review tomorrow.

SiegeLord says it affects other formats than just WAV too, so I'm wondering if the glitch may not be in allegro-acodec but in the main audio code.

**Comment by SiegeLord** [comment]
I'll try looking again today as well. This is next on my list :godmode:.

**Comment by fatcerberus** [comment]
Hm, I may have a lead.  In the stream feeder thread:

```c
        /* In case it reaches the end of the stream source, stream feeder will
         * fill the remaining space with silence. If we should loop, rewind the
         * stream and override the silence with the beginning.
         * In extreme cases we need to repeat it multiple times.
         */
         while (bytes_written < bytes &&
                  stream->spl.loop == _ALLEGRO_PLAYMODE_STREAM_ONEDIR) {
            size_t bw;
            al_rewind_audio_stream(stream);
            maybe_lock_mutex(stream->spl.mutex);
            bw = stream->feeder(stream, fragment + bytes_written,
               bytes - bytes_written);
            bytes_written += bw;
            maybe_unlock_mutex(stream->spl.mutex);
         }
```

The comment says it fills leftover space with silence, but I don't see any place where that would happen (there is no `al_fill_silence()` call).  The individual feeder functions in the acodec addon, e.g. `wav_stream_update()` don't seem to do it either.

**Comment by SiegeLord** [comment]
Alright, so I think I probably have fixed it as of 3919ac2. There were a few issues:

1. Pulseaudio backend was corrupting the end of samples, which proved to be a roadblock in that original thread. I managed to fix it.
2. Indeed, adding `al_fill_silence` call fixed the static, but the `bleep.wav` from the original thread still had a click at the end...
3. Turns out, `bleep.wav` has a click in the file itself. If you smooth it out, then the click goes away.

**Comment by fatcerberus** [comment]
Thanks, I'll test this out tomorrow.

**Comment by fatcerberus** [comment]
Hm, there's still an audible "pop" at the end of WAV streams that isn't present for the .ogg equivalents.  I'll test with ex_acodec and see what happens there.

**Comment by fatcerberus** [comment]
ex_acodec exhibits the same issue, the reverb effect masks it somewhat but there's definitely a difference between WAV and ogg output that can't simply be chalked up to compression artifacts.  Platform is Windows, w/ DSound backend.

**Comment by SiegeLord** [comment]
Would you mind recording it via audacity? That's how I've been diagnosing this. On Linux, I now get identical waveforms between what's on disk and what comes out of the speakers. I fully accept the possibility that DSound backend might be broken like Pulseaudio was. I'll start testing this out on Windows today.

**Comment by fatcerberus** [comment]
Audacity file created by recording from Stereo Mix:
https://drive.google.com/open?id=0BxPKLRqQOUSNNWg1LTg3dzA5QkU

The same sound is played twice, first from a .wav file and then from a transcoded .ogg file.  Comparing the waveforms, there's definitely some kind of corruption at the end of the wav output.

**Comment by fatcerberus** [comment]
Ignore that link, Audacity screwed up.  Let me try again...

**Comment by fatcerberus** [comment]
Here, I just exported the project as a WAV file:
https://drive.google.com/open?id=0BxPKLRqQOUSNUnZiRFJKVkl0Rm8

**Comment by SiegeLord** [comment]
Thanks. I managed to reproduce it both on Linux and Windows with `heal.wav`. The issue turns out to be that when streaming wav files, we're possibly reading past the end of the audio data chunk, and outputting the rest of the file as audio. I haven't implemented a fix (it took me too long to discover that), but will do so tomorrow.

Note that this doesn't explain the `ex_acodec` difference you're hearing. I couldn't reproduce that, and that uses a different codepath. I verified that at least for `al_load_sample`, both ogg and wav load the very same data. I'd encourage you to try to verify that difference using audacity too.

I'll submit a fix for the issue I found tomorrow and hopefully it'll be gone for good.

**Comment by fatcerberus** [comment]
So in other words, there's garbage data at the end of `heal.wav`?  I noticed in the wav plugin that the streamer just `fread`s until it gets to the end of the file but I didn't think anything of it because I assumed that was just how the format was set up.

As for `ex_acodec`, I'll test that again, maybe it was just a placebo effect, or perhaps the reverb is amplifying compression artifacts.

**Comment by fatcerberus** [comment]
You're right, `ex_acodec` is fine.  The ogg waveforms are more "messy" due to compression, but otherwise I don't see any corruption.  It definitely sounds different though... so yeah, I guess the reverb is just making the compression more obvious similar to how upscaling a video highlights all the artifacts.

**Comment by SiegeLord** [comment]
Alright, this commit implements the change I had in mind: 2a75bec97e56afa7d2fab299d294b0cea7393892

Please try it and hopefully that's the end of this.

**Comment by fatcerberus** [comment]
That did it, thanks!  Good to finally have this fixed!  :smile:

**Comment by SiegeLord** [comment]
Great!
